modules/ds18b20/ds18b20.c

```diff
@@ -0,0 +1,467 @@
+/** 
+ * |----------------------------------------------------------------------
+ * | Copyright (c) 2016 Tilen MAJERLE
+ * |  
+ * | Permission is hereby granted, free of charge, to any person
+ * | obtaining a copy of this software and associated documentation
+ * | files (the "Software"), to deal in the Software without restriction,
+ * | including without limitation the rights to use, copy, modify, merge,
+ * | publish, distribute, sublicense, and/or sell copies of the Software, 
+ * | and to permit persons to whom the Software is furnished to do so, 
+ * | subject to the following conditions:
+ * | 
+ * | The above copyright notice and this permission notice shall be
+ * | included in all copies or substantial portions of the Software.
+ * | 
+ * | THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * | EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * | OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE
+ * | AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * | HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * | WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+ * | FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * | OTHER DEALINGS IN THE SOFTWARE.
+ * |----------------------------------------------------------------------
+ */
+#include "ds18b20.h"
+
+/**
+ * @brief  Private function converts config register value to uint8_t resolution
+ * @param  config_register: value of configuration register.
+ * @return Resolution in bits: 9, 10, 11, or 12 
+ */
+uint8_t DS18B20_ConfigRegToResolution(uint8_t config_register) {
+    return ((config_register & DS18B20_CONFIG_REGISTER_R0_R1_BITMASK) 
+        >> DS18B20_CONFIG_REGISTER_RESERVED_BITS) + 9;
+}
+
+DS18B20_Status DS18B20_Start(OneWire_t* OneWire, uint8_t *ROM) {
+    if (!OneWire || !ROM) return DS18B20_USAGE_ERROR;
+    if (DS18B20_Is(ROM)==DEVICE_NOT_DS18B20) return DEVICE_NOT_DS18B20;
+    
+    /* Reset line */
+    if (OneWire_Reset(OneWire) == ONEWIRE_FAILURE) return DS18B20_FAILURE;
+    /* Select ROM number */
+    OneWire_SelectWithPointer(OneWire, ROM);
+    /* Start temperature conversion */
+    OneWire_WriteByte(OneWire, DS18B20_CMD_CONVERTTEMP);
+    
+    return DS18B20_SUCCESS;
+}
+
+DS18B20_Status DS18B20_StartAll(OneWire_t* OneWire) {
+    if (!OneWire) return DS18B20_USAGE_ERROR;
+    /* Reset pulse, abort if presence pulse is not good */
+    if (OneWire_Reset(OneWire) == ONEWIRE_FAILURE) return DS18B20_FAILURE;
+    /* Skip rom */
+    OneWire_WriteByte(OneWire, ONEWIRE_CMD_SKIPROM);
+    /* Start conversion on all connected devices */
+    OneWire_WriteByte(OneWire, DS18B20_CMD_CONVERTTEMP);
+    return DS18B20_SUCCESS;
+}
+
+DS18B20_Status DS18B20_Read(OneWire_t* OneWire, uint8_t *ROM, float *destination) {

```

Comment:
I'm just curious...  since this read function is a blocking function, have you benchmarked how long it takes to complete this operation?  That info may come in handy later when debugging any issues.

Reply:
Same question for the write function

Reply:
Along the same lines of questions as Robbie's, since this is bit-banging onewire in a preemptive RTOS, do we know how much leeway we have in the bit timings for onewire?